### DISCUSSION TITLE
USB Serial Write

### DISCUSSION THREAD

**Discussion by AThilenius** [mciantyre/teensy4-rs#143] [Q&A]
Is there a "simple" way to write bytes to the USB as a serial device? I want to transmit IMU data to a desktop application. I could live with the logging API for this if un-framed binary data can be sent with it? I see examples out there for interacting with USB via the HAL APIs, but they are all monstrously complicated and hurt my Grug brain. At the risk of sounding like a total luddite, I'm hoping for something akin to the Adruino `Serial.write()` 🤞

Btw... writing Rust on the Teensy is a dream come true, so thank you ❤️

**Comment by mciantyre** [comment]
How about using [defmt](https://defmt.ferrous-systems.com/introduction.html) to approximate `Serial.write()`? I think defmt supports framing in its encoding, and it's already supported by the USB serial logger. This approach is simple in firmware, but it may be less simple on the host. 8983626 shows a demo.

**Comment by AThilenius** [comment]
It's an option, but seems to still be really logging-centric, carrying a lot of metadata like timestamp, level, location, file, line et. al. I would actually prefer unframed (ie. just a byte stream) because I'll be framing messages myself. I'm hoping this isn't that uncommon a thing to want to do.

**Reply by mciantyre** [reply]
I'm sorry, I forgot to follow up. We don't have anything easier than the logging interfaces in any of today's crates. You could take an approach [like this](https://github.com/mciantyre/imxrt-hal/commit/319a5814df4ff964f47ccba33cfee82e52c1ad72) ([demo](https://github.com/mciantyre/teensy4-rs/commit/52d4de0b5b5e1a7ec43a436be1c1e705f7aa542d)), but I'm not sure I'd like to support that API in `imxrt-log`.

**Comment by stappersg** [comment]
On Thu, Jul 20, 2023 at 08:53:49AM -0700, Alec Thilenius wrote:
> It's an option,

Acknowledge
